File: strings/word-subsets/word-subsets.go
package wordsubsets

func wordSubsets(A []string, B []string) []string {
	maxFreqB := [26]int{}

	for _, b := range B {
		freqB := count(b)

		for i := range maxFreqB {
			if freqB[i] > maxFreqB[i] {
				maxFreqB[i] = freqB[i]
			}
		}
	}

	ans := make([]string, 0)

	for i := range A {
		freqA := count(A[i])

		if compare(freqA, maxFreqB) {
			ans = append(ans, A[i])
		}
	}

	return ans
}

func count(S string) [26]int {
	freq := [26]int{}

	for i := range S {
		freq[S[i]-'a']++
	}

	return freq
}

func compare(A [26]int, B [26]int) bool {
	for i := range A {
		if A[i] < B[i] {
			return false
		}

		return true
	}
}

File: strings/word-subsets/word-subsets.java
class Solution {
  public List<String> wordSubsets(String[] A, String[] B) {
    int[] maxFreqB = new int[26];
    for (String b : B) {
      int[] freqB = count(b);

      for (int i = 0; i < 26; i++) {
        maxFreqB[i] = Math.max(maxFreqB[i], freqB[i]);
      }
    }

    List<String> ans = new ArrayList();

    search: for (String a : A) {
      int[] freqA = count(a);
      if (compare(freqA, maxFreqB)) {
        ans.add(a);
      }
    }

    return ans;
  }

  public int[] count(String S) {
    int[] freq = new int[26];

    for (char c : S.toCharArray()) {
      freq[c - 'a']++;
    }
    return freq;
  }

  public boolean compare(int[] A, int[] B) {
    for (int i = 0; i < 26; i++) {
      if (A[i] < B[i]) {
        return false;
      }
    }

    return true;
  }
}
